### PR TITLE
rgw: add support for Swift's TempURLs with prefix-based scope

### PR DESCRIPTION
**Pull request by rzarzynski**

### CODE REVIEW

File: src/rgw/rgw_string.h [hunk 1]
```diff
@@ -222,5 +222,9 @@ std::string string_join_reserve(char delim, const Args&... args)
 {
   return string_join_reserve(boost::string_view{&delim, 1}, args...);
 }
-
+template <typename... Args>
+std::string string_concat_reserve(const Args&... args)
+{
+  return string_join_reserve(boost::string_view{}, args...);
+}
```
Comment (cbodley) [inline src/rgw/rgw_string.h]: see `string_cat_reserve()` above

Comment (rzarzynski) [inline src/rgw/rgw_string.h]: I've completely missed it out. Thanks, @cbodley!

File: src/rgw/rgw_swift_auth.cc [hunk 2]
```diff
@@ -190,20 +192,76 @@ class TempURLEngine::SignatureHelper
...
+    }
+  }
+
+  bool is_equal_to(const std::string& rhs) const {
+    bool is_auth_ok = base_t::is_equal_to(rhs);
+
+    if (prefix) {
+      const auto prefix_uri = string_cat_reserve(no_obj_uri, *prefix);
+      is_auth_ok = boost::algorithm::starts_with(decoded_uri, prefix_uri);
+    }
```
Comment (cbodley) [inline src/rgw/rgw_swift_auth.cc]: i'm having trouble following this logic. does this prefix case really not depend on the input string `rhs`?

**Comment by rzarzynski** [comment]
@cbodley: after fixing unrelated issues in TempURL (#16636, #16658, #16659), the `TestTempURLPrefix` became green:
```
(swift) [rzarzynski@rzarzynski functional]$ SWIFT_TEST_CONFIG_FILE=rgw.conf nosetests -v test.functional.test_tempurl:TestTempURLPrefix
SKIPPING FUNCTIONAL TESTS SPECIFIC TO AUTH VERSION 3
SKIPPING FUNCTIONAL TESTS SPECIFIC TO SERVICE TOKENS
SKIPPING FUNCTIONAL TESTS DUE TO NO CONFIG FOR RESELLER ADMIN
SKIPPING FUNCTIONAL TESTS SPECIFIC TO AUTH VERSION 3
SKIPPING FUNCTIONAL TESTS SPECIFIC TO SERVICE TOKENS
SKIPPING FUNCTIONAL TESTS DUE TO NO CONFIG FOR RESELLER ADMIN
test_GET (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_GET_DLO_inside_container (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_GET_DLO_outside_container (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_GET_with_key_2 (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_HEAD (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_PUT (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_PUT_manifest_access (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_changing_expires (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_changing_sig (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_different_object (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_empty_prefix (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_missing_query_parm (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_no_prefix_match (test.functional.test_tempurl.TestTempURLPrefix) ... ok
test_object_url_with_prefix (test.functional.test_tempurl.TestTempURLPrefix) ... ok

----------------------------------------------------------------------
Ran 14 tests in 5.403s

OK
```

**Comment by rzarzynski** [comment]
I'm setting following configurables:
```
rgw_swift_enforce_content_length=true
rgw_keystone_implicit_tenants=true
rgw_keystone_accepted_admin_roles "admin"
rgw_swift_versioning_enabled=true
# especially following ones are crucial
rgw_swift_account_in_url=true
rgw_swift_auth_entry="swift/auth" 
rgw frontends = civetweb port=8000 prefix=/swift
```

Also we need to workaround the lack of frontend's `prefix` support in `TempAUTH` as the whole Swift API must be placed in the root of URL hierarchy:
```
diff --git a/src/rgw/rgw_swift_auth.cc b/src/rgw/rgw_swift_auth.cc
index e7a25f5..e64c73a 100644
--- a/src/rgw/rgw_swift_auth.cc
+++ b/src/rgw/rgw_swift_auth.cc
@@ -695,7 +695,7 @@ void RGW_SWIFT_Auth_Get::execute()
     tenant_path.append(info.user_id.to_str());
   }
 
-  dump_header(s, "X-Storage-Url", swift_url + swift_prefix + "/v1" +
+  dump_header(s, "X-Storage-Url", swift_url + "/v1" +
               tenant_path);
 
   using rgw::auth::swift::encode_token;
```

**Comment by yuriw** [comment]
http://pulpito.ceph.com/yuriw-2017-08-01_19:45:33-rgw-wip-yuri-testing3_2017_8_1_2-distro-basic-smithi/

**Pull request merged**